### PR TITLE
fix: default value reflects the runtime environment variables rather than the ones present at import time.

### PR DESCRIPTION
**Pull request by junrae6454**
# What does this PR do?
Refactor deterministic flag handling in Flash Attention forward method

- Set the `deterministic` flag to False by default in `_flash_attention_forward`.
- Retrieve the deterministic flag from the environment variable `FLASH_ATTENTION_DETERMINISTIC` within the function scope to recognize runtime environment variables.
- Adjusted `flash_kwargs` dictionary to use the updated deterministic flag handling mechanism.

This change ensures that the parameter's default value reflects the runtime environment variables rather than the ones present at import time.

```python
>>> import os
>>> def default_param(param = os.environ.get("deterministic", "0") == "1"):
...     print(param)
...
>>> default_param()
False
>>> os.environ["deterministic"] = "1"
>>> default_param()
False
```
```python
>>> import os
>>> os.environ["deterministic"] = "1"
>>> def default_param(param = os.environ.get("deterministic", "0") == "1"):
...     print(param)
...
>>> default_param()
True
>>> os.environ["deterministic"] = "0"
>>> default_param()
True
```

Reference: https://github.com/huggingface/transformers/pull/31961

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Thanks to @ArthurZucker

### CODE REVIEW

File: src/transformers/modeling_flash_attention_utils.py [hunk 1]
```diff
@@ -182,7 +182,8 @@ def _flash_attention_forward(
     flash_kwargs = {"window_size": (sliding_window, sliding_window)} if use_sliding_windows else {}
 
     if is_flash_attn_greater_or_equal("2.4.1"):
-        flash_kwargs["deterministic"] = deterministic
+        is_deterministic = os.environ.get("FLASH_ATTENTION_DETERMINISTIC", "0") == "1"
+        flash_kwargs["deterministic"] = is_deterministic
```
Comment (amyeroberts) [inline src/transformers/modeling_flash_attention_utils.py:186]: If this is the case, then the `deterministic` flag has no effect. It should either be accounted for or removed

**Review by amyeroberts — COMMENTED** [review]
Hi @junrae6454, thanks for opening a PR! 

Eeek - yes, having the function use the env var as a default is really bad. I'd suggest having `deterministic` default to `None` and use the env var if it's unset

File: src/transformers/modeling_flash_attention_utils.py (hunk 1 above)
Comment (junrae6454) [reply src/transformers/modeling_flash_attention_utils.py:186]:
```suggestion
    if is_flash_attn_greater_or_equal("2.4.1"):
        if deterministic is None:
            deterministic = os.environ.get("FLASH_ATTENTION_DETERMINISTIC", "0") == "1"
        flash_kwargs["deterministic"] = deterministic
```

Would it be better to change the default value of the deterministic option to None and modify the code as shown above?

Comment (amyeroberts) [reply src/transformers/modeling_flash_attention_utils.py:186]: Yep - LGTM!

Comment (junrae6454) [reply src/transformers/modeling_flash_attention_utils.py]: Thank you so much, @amyeroberts , for your help with this pull request!

File: src/transformers/modeling_flash_attention_utils.py [hunk 2]
```diff
@@ -142,7 +142,7 @@ def _flash_attention_forward(
     sliding_window: Optional[int] = None,
     use_top_left_mask: bool = False,
     softcap: Optional[float] = None,
-    deterministic: bool = os.environ.get("FLASH_ATTENTION_DETERMINISTIC", "0") == "1",
+    deterministic: bool = None,
```
Comment (amyeroberts) [inline src/transformers/modeling_flash_attention_utils.py:145]: nit

```suggestion
    deterministic: Optional[bool] = None,
```

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!

**Comment by amyeroberts** [comment]
For the quality checks, could you rebase on main to include the upstream changes. This should fix things

**Pull request merged**